ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/Util/HardFork/OracularClock.hs

```diff
@@ -0,0 +1,242 @@
+               -> (a -> m ())
+               -> m ()
+withinEachSlot clock s0 get put = do
+    (s1, mbA) <- withinTheCurrentSlot clock s0 get
+    case mbA of
+      Nothing ->
+          -- ticked while we were trying, so try again in this new slot
+          withinEachSlot clock s1 get put
+      Just a  -> do
+          put a
+          -- we succeed in this slot, so try again in the next
+          let s2 = succ s1
+          void $ blockUntilSlot clock s2
+          withinEachSlot clock s2 get put
+
+-- | See 'OracularClock'
+new :: forall m. (IOLike m, HasCallStack)
+    => BTime.SystemTime m
+    -> ResourceRegistry m
+    -> NumSlots
+    -> Future
+    -> m (OracularClock m)
+new systemTime@BTime.SystemTime{..} registry (NumSlots n) future = do
+    waitOneSlot <- do
+      slotLengthsVar <- uncheckedNewTVarM $ futureSlotLengths future
+      pure $ do
+        slotLength <- atomically $ do
+          x :< xs <- readTVar slotLengthsVar
+          x <$ writeTVar slotLengthsVar xs
+        threadDelay $ nominalDelay $ BTime.getSlotLength slotLength
+    let _ = waitOneSlot :: m ()
+
+    slotVar <- newTVarM 0
+    doneVar <- newTVarM False
+    void $ forkThread registry "OracularClock.ticker" $ do
+      -- The first slot has already begun, so let other threads execute until
+      -- the slot ends
+      waitOneSlot
+      -- And repeat for the @n - 1@ remaining slots
+      replicateM_ (fromIntegral n - 1) $ do
+        atomically $ modifyTVar slotVar (+ 1)
+        waitOneSlot
+      -- Set internal flag indicating the clock is exhausted (ie the test is
+      -- over)
+      atomically $ writeTVar doneVar True
+
+    let readCurrentSlotSTM = do
+            -- block forever if clock is exhausted
+            readTVar doneVar >>= check . not
+
+            readTVar slotVar
+
+    let finiteSystemTimeCurrent = do
+            t <- systemTimeCurrent
+
+            -- block forever if clock is exhausted (this checks the time
+            -- directly, without relying on 'doneVar')
+            atomically $ do
+              let totalDelta =
+                      (sum . map BTime.getSlotLength) $
+                      (take (fromIntegral n) . toList) $
+                      futureSlotLengths future
+                  tFinal = BTime.RelativeTime totalDelta
+              check $ t < tFinal
+
+            pure t
+
+    let getPresent = do
+            BTime.RelativeTime t <- finiteSystemTimeCurrent
+            pure $ futureTimeToSlot future t
+
+    pure OracularClock
+      { blockUntilSlot = \slot -> atomically $ do
+          now <- readCurrentSlotSTM
+          case compare now slot of
+            LT -> retry
+            EQ -> pure False
+            GT -> pure True   -- ie " too late "
+      , delayUntilNextSlot = do
+          (_slot, inSlot, slotLength) <- getPresent
+          pure $ BTime.getSlotLength slotLength - inSlot
+      , finiteSystemTime = systemTime
+          { BTime.systemTimeCurrent = finiteSystemTimeCurrent
```

Comment:
I am concerned about this. This is the `SystemTime` that in the end gets passed to the main consensus infrastructure (for the leadership check and for the in-future check in the chain DB), but unlike a "real" system clock, this one has infinite look-ahead. It would be better to limit it by a safe zone to increase the strength of the tests. That doesn't necessarily have to be done as part of this PR, but if not, we should create a ticket and do it relatively soon.

Reply:
I don't know what you mean by "look-ahead" in this context.

https://github.com/input-output-hk/ouroboros-network/blob/6507ec6eb08e75925e9c3072a11db84e226bf706/ouroboros-consensus/src/Ouroboros/Consensus/BlockchainTime/WallClock/Types.hs#L77-L90

Reply:
Oh! Apologies. Yes, no, you are right. I was thinking of the `BlockchainTime`. As long as the `BlockchainTime` that we pass to `InFuture` and the ladershipcheck can't look ahead indefinitely, we're good.